mlflow/store/sqlalchemy_store.py

```diff
@@ -184,48 +186,68 @@ def log_metric(self, run_uuid, metric):
             raise MlflowException('Metric={} must be unique'.format(metric),
                                   INVALID_PARAMETER_VALUE)
 
+    def get_metric(self, run_uuid, metric_key):
+        metric = self.session.query(SqlMetric).filter_by(
+            run_uuid=run_uuid, key=metric_key
+        ).order_by(sqlalchemy.desc(SqlMetric.timestamp)).first()
+
+        if metric is None:
+            raise MlflowException('Metric={} does not exist'.format(metric_key),
+                                  RESOURCE_DOES_NOT_EXIST)
+
+        return metric.to_mlflow_entity()
+
+    def get_metric_history(self, run_uuid, metric_key):
+        metrics = self.session.query(SqlMetric).filter_by(run_uuid=run_uuid, key=metric_key).all()
+        return [metric.to_mlflow_entity() for metric in metrics]
+
+    def get_all_metrics(self, run_uuid):
+        max_ts = label("timestamp", func.max(SqlMetric.timestamp))
+        subq = self.session.query(SqlMetric.key, max_ts)\
+            .filter_by(run_uuid=run_uuid)\
+            .group_by(SqlMetric.key)\
+            .subquery()
+        metrics = self.session.query(SqlMetric)\
+            .filter_by(run_uuid=run_uuid)\
+            .join(subq, and_(SqlMetric.key == subq.c.key,
+                             SqlMetric.timestamp == subq.c.timestamp)).all()
+        return [metric.to_mlflow_entity() for metric in metrics]
+
     def log_param(self, run_uuid, param):
         # if we try to update the value of an existing param this will fail
         # because it will try to create it with same run_uuid, param key
         try:
             self._get_or_create(SqlParam, run_uuid=run_uuid, key=param.key,
                                 value=param.value)
         except sqlalchemy.exc.IntegrityError:
-            raise MlflowException('changing parameter {} value is not allowed'.format((run_uuid,
-                                                                                      param)),
+            raise MlflowException('Changing param value is not allowed. Run={}'.format(run_uuid),
```

Comment:
Including the parameter key is probably not a bad idea (and maybe the value too)?